### ISSUE TITLE
Missing Conformer Module in CRNN Code

### ISSUE THREAD

**Issue by LJF-SAL** [oldtown11/sound-event-detection#1]
Hi there! 👋

I noticed that in the CRNN part of the code, there is an import referencing:

`from .conformer.conformer_encoder import ConformerEncoder, ConformerPromptedEncoder  `
However, it seems the conformer module is not included in the current repository. Could you please provide the missing Conformer-related files or update the repository to include them?

Thanks for your help — looking forward to your update!

**Comment by oldtown11** [comment]
The `conformer_encoder` package retained here was inadvertently not removed, but it was not used in the actual training process.

Initially, we attempted to replace the RNN in the CRNN with a Conformer structure. Experimental results showed that the Conformer provided significant improvement on the classification task (i.e., higher PSDS2 score), but only limited gain on the localization task (relatively smaller improvement in PSDS1).

Considering the overall balance between localization and classification performance, we ultimately chose to retain the baseline RNN structure, which performs steadily in both tasks.

Based on previous experimental results:
- For the localization metric (PSDS1): CRNN > Conformer
- For the classification metric (PSDS2): CRNN < Conformer

**Comment by oldtown11** [comment]
I'm very sorry. I'm currently engaged in other work, so the code organization is a bit messy.

If you have any questions, please feel free to ask!

**Comment by LJF-SAL** [comment]
> The `conformer_encoder` package retained here was inadvertently not removed, but it was not used in the actual training process.
> 
> Initially, we attempted to replace the RNN in the CRNN with a Conformer structure. Experimental results showed that the Conformer provided significant improvement on the classification task (i.e., higher PSDS2 score), but only limited gain on the localization task (relatively smaller improvement in PSDS1).
> 
> Considering the overall balance between localization and classification performance, we ultimately chose to retain the baseline RNN structure, which performs steadily in both tasks.
> 
> Based on previous experimental results:
> 
> * For the localization metric (PSDS1): CRNN > Conformer
> * For the classification metric (PSDS2): CRNN < Conformer

Thank you so much for your previous prompt reply and assistance!

Following your guidance, we have successfully integrated the ffine-tuning BEATs code you provided into the DCASE 2024 Task 4 baseline system. As instructed, we commented out the unused Conformer-related import (from .conformer.conformer_encoder import ConformerEncoder, ConformerPromptedEncoder) and ran the complete experiment.

We would like to share our initial experimental results:

PSDS1: 0.3803

mPAUC: 0.6868

We noticed that these results, particularly the PSDS1 score, show a noticeable gap compared to the strong performance reported in your paper. We speculate that the discrepancy might be attributed to several factors:

Hyperparameter Tuning: Specific optimizations might have been made to the original baseline's hyperparameters (e.g., learning rate, weight decay, batch size, number of epochs).

Model Initialization: The CRNN model might have been initialized with weights pre-trained using the frozen BEATs encoder before fine-tuning.

Training Strategy: Different learning rate schedulers, data augmentation mixes, or other advanced training techniques might have been employed.

We greatly appreciate you open-sourcing your code and providing support. Your work and this codebase have been immensely helpful and provided valuable insights. If you have any thoughts on this performance difference or could hint at any critical configuration details we might have overlooked, we would be very grateful.

Thank you again for your time and contribution!

Best regards

**Comment by oldtown11** [comment]
Thank you very much for your attention! Last week, I was still conducting ablation experiments for the DCASE2024 task, and the results obtained using only the Prompt module are generally consistent with the performance reported in the paper.

Our Prompt module is designed to be plug-and-play. For this task, all other modules and hyperparameter settings remain the same as the baseline method. Before identifying the cause of the issue, I suggest you comment out the Prompt section to test whether the baseline performance can be reproduced.

The problem you encountered is likely due to some oversight in my open-source code, for which I sincerely apologize. I am currently away from campus, but I expect to conduct a comprehensive review of the code and address the issues as soon as possible over the weekend.

**Comment by LJF-SAL** [comment]
Hello again,


Thank you for your previous replies and support. After further examination and experimentation, I have identified a key point regarding the performance gap.


I noticed that the currently provided CRNN code in the repository is adapted for the DCASE 2023 baseline. However, the official DCASE 2024 baseline introduced a significant change: the incorporation of the soft-labeled dataset. This introduction necessitated modifications to the CRNN model to handle these soft labels effectively.


Therefore, could you please upload the version of the CRNN code that is specifically adapted for the DCASE 2024 baseline, including the necessary changes to support the soft-labeled dataset?


Thank you for your time and consideration!

**Comment by oldtown11** [comment]
Hi ! I'm sorry for the delayed response.

Last week, I re-evaluated the method for the 24 dataset in the paper, and the results are basically consistent with the performance reported in the paper. Here, I have conducted a comprehensive analysis of the issue you encountered.

<img width="690" height="268" alt="Image" src="https://github.com/user-attachments/assets/ee7fdcd0-17d0-481b-88fe-b98b8431af11" />

1. First, an explanation regarding your issue:

   - I did make some adjustments (including data augmentation) to the DCASE23 baseline. However, for the 24 dataset, no other modifications were made except for migrating our new method to the DCASE24 baseline. Therefore, I believe the score drop you encountered may not be an issue with the model or the data itself.

2. Based on the above, I analyze that the discrepancy may be caused by the following reasons:

   - Training cycle setting: I noticed that the `n_epochs`  of the baseline system can lead to severe overfitting of PSDS1. Experiments show that as the number of training epochs increases, the `PSDS1` metric exhibits obvious overfitting (for example, `mp_auc` can reach 0.78, while `PSDS1` is only about 0.45). Therefore, in this experiment, I set `n_epochs = 140`.

   - Difference in test set usage: I used the public test set (public set) for evaluation, rather than the validation_16k dataset. This may be the main reason for the score discrepancy. If you need this public test set, I can provide it to you.

3. Suggestions for some parameters/metrics
   - Model saving strategy: It is recommended that you set the `save_top_k` parameter to 10 or higher in train_pretrained.py. This allows retaining more intermediate models for performance evaluation, thereby selecting the optimal model (in our experiments, we found that some high-performing models would be overwritten).
